aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -253,6 +251,259 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
+    tensors[0] = _tensors[0].unsqueeze(0);
+  } else if (_tensors[0].dim() == 2) {
+    tensors[0] = _tensors[0];
+    out_shape.push_back(tensors[0].size(0));
+  } else {
+    TORCH_CHECK(
+        false,
+        "multi_dot(): the first tensor must be 1D or 2D but got ",
+        _tensors[0].dim(),
+        "D");
+  }
+
+  // If the last tensor is 1D of size n view it as a column vector (n, 1)
+  if (_tensors[n - 1].dim() == 1) {
+    tensors[n - 1] = _tensors[n - 1].unsqueeze(-1);
+  } else if (_tensors[n - 1].dim() == 2) {
+    tensors[n - 1] = _tensors[n - 1];
+    out_shape.push_back(tensors[n - 1].size(1));
+  } else {
+    TORCH_CHECK(
+        false,
+        "multi_dot(): the last tensor must be 1D or 2D but got ",
+        _tensors[0].dim(),
+        "D");
+  }
+
+  // Ensure middle tensors are 2D
+  for (const auto i : c10::irange<size_t>(1, n - 1)) {
+    TORCH_CHECK(
+        _tensors[i].dim() == 2,
+        "multi_dot(): tensor ",
+        i,
+        " must be 2D but got ",
+        _tensors[0].dim(),
+        "D");
+    tensors[i] = _tensors[i];
+  }
+
+  // Ensure all tensors have the same device and dtype and check
+  // that the shapes can be multiplied
+  const auto dtype = tensors[0].dtype();
+  const auto device = tensors[0].device();
+  bool has_empty_tensor = tensors[0].numel() == 0;
+  for (const auto i : c10::irange<size_t>(1, n)) {
+    TORCH_CHECK(
+        tensors[i].dtype() == dtype,
+        "multi_dot(): all tensors must have be the same dtype but tensor 0 is ",
+        dtype,
+        " and tensor ",
+        i,
+        " ",
+        tensors[i].dtype());
+    TORCH_CHECK(
+        tensors[i].device() == device,
+        "multi_dot(): all tensors must be on the same device but tensor 0 is on ",
+        device,
+        " and tensor ",
+        i,
+        " on ",
+        tensors[i].device());
+    TORCH_CHECK(
+        tensors[i - 1].size(-1) == tensors[i].size(0),
+        "multi_dot(): tensors ",
+        i - 1,
+        " and ",
+        i,
+        " with shapes ",
+        _tensors[i - 1].sizes(),
+        " and ",
+        _tensors[i].sizes(),
+        " cannot be multiplied")
+    has_empty_tensor |= tensors[i].numel() == 0;
+  }
+
+  Tensor result;
+
+  if (out.defined()) {
+    TORCH_CHECK(
+        dtype == out.dtype(),
+        "multi_dot(): expected out tensor to have dtype ",
+        dtype,
+        " but got ",
+        out.dtype());
+    TORCH_CHECK(
+        device == out.device(),
+        "multi_dot(): expected out tensor to be on device ",
+        device,
+        " but got ",
+        out.device());
+
+    at::native::resize_output(out, out_shape);
+
+    // View output as 2D for computations
+    result = out.view({tensors[0].size(0), tensors.back().size(-1)});
+  }
+
+  // The resize_ and view calls below are to ensure the
+  // output shape respects the original dimensionality of
+  // the first and last tensors which we are now viewed as 2D
+
+  if (has_empty_tensor) {
```

Comment:
How can this happen? Wouldn't every tensor in the computation have to be empty for this to occur?

Reply:
If at least one tensor is empty, then it has a zero size dimension. When multiplying this tensor if the size zero dimension is an inner dimension the result is a tensor of zeros, if it's an outer dimension the result is another empty tensor. Since tensors can be multiplied in any order, it follows by induction that the result of multiplying all the tensors will be either a tensor of zeros or an empty tensor. (handwaving proof here)

Reply:
I got it ;)